### PR TITLE
Bugfix: ensure meson gets correct compiler flags

### PR DESCRIPTION
**Pull request by SSE4**
closes: https://github.com/conan-io/conan/issues/4573
tags: @slow, @svn
Changelog: Bugfix: Meson build-helper gets correct compiler flags, AutoTools build environment adds compiler.runtime flags
Docs: omit

short story: currently meson build helper doesn't propagate required compiler flags, like `AutoToolsBuildEnvironment`, `CMake` or `MSBuild` do. it requires ugly workarounds in each recipe:
```
        if str(self.settings.compiler) in ["gcc", "clang"]:
            if self.settings.arch == "x86":
                defs["c_args"] = "-m32"
                defs["cpp_args"] = "-m32"
                defs["c_link_args"] = "-m32"
                defs["cpp_link_args"] = "-m32"
            elif self.settings.arch == "x86_64":
                defs["c_args"] = "-m64"
                defs["cpp_args"] = "-m64"
                defs["c_link_args"] = "-m64"
                defs["cpp_link_args"] = "-m64"
        elif self.settings.compiler == "Visual Studio":
            defs["c_args"] = "-%s" % self.settings.compiler.runtime
            defs["cpp_args"] = "-%s" % self.settings.compiler.runtime
```
as meson supports the same way as auto-tools, I've re-used `AutoToolsBuildEnvironment` build helper in order to set up correct compile flags.
see also: https://mesonbuild.com/howtox.html#set-extra-compiler-and-linker-flags-from-the-outside-when-eg-building-distro-packages

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/build/meson.py [hunk 1]
```diff
@@ -151,15 +152,19 @@ def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
...
-            command = "%s && %s" % (vcvars, command)
-        return command
+    @property
+    def _vcvars_needed(self):
+        return self._compiler == "Visual Studio" and self.backend == "ninja"
+
+    def _run(self, command):
+        with tools.vcvars(self._settings,
+                          output=self._conanfile.output) if self._vcvars_needed else tools.no_op():
+            env_build = AutoToolsBuildEnvironment(self._conanfile)
```
Comment (lasote) [inline conans/client/build/meson.py]: At first sight it doesn't feel great to use another build helper and especially that one based only on the env vars.
Is there another thing we can do like write a file to be injected to meson?

**Comment by jgsogo** [comment]
I would have a look at the _Meson build description_ (`meson.build` files) approach, although I'm not sure how can we plug it into an existing project 🤔

**Comment by SSE4** [comment]
this approach may work: https://mesonbuild.com/Cross-compilation.html
I'll take a look into it

File: conans/client/build/meson.py (hunk 1 above)
Comment (SSE4) [reply conans/client/build/meson.py]: seems like writing to file is something we will need to solve https://github.com/conan-io/conan/issues/4529
but I think it's blocked by https://github.com/conan-io/conan/pull/5184 
we need to do some design decisions on how to handle such kind of files, are they produced by their own generators, or by existing generators, or some other entities.
I think right now it's too much for the original particular problem.
I prefer to go by small iterations, like first solve flags issue, then solve cross-compiling issue, and then solve toolchain file design problems. I think it's too much to do at single shot.
for the `AutoToolsBuildEnvironment`, why not re-use it, at least for the first iteration? as soon as meson is compatible with auto-tools variables, we can re-use the code, especially as it's already well-tested and proven to work correctly.

**Comment by lasote** [comment]
I'm very confuse about the new generator for cross building with Meson that you introduced. [Here](https://github.com/conan-io/conan/pull/5222#discussion_r288121174) you are suggesting to go step by step and then suddenly invent a new generator?
Unless you can explain clearly why is that needed for closing the issue I would ask you to create an issue and a different pull request, that won't be a priority until we review the full picture of cross building.

**Comment by SSE4** [comment]
@lasote I am going to remove that generator from the PR, it's still incomplete,

**Comment by SSE4** [comment]
generator code was removed. it will go to its own PR, probably after the https://github.com/conan-io/conan/pull/5184

**Comment by jgsogo** [comment]
I agree with adding all those flags to Meson build helper, they are needed and very useful. I also like the implementation on top of `AutoToolsBuildEnvironment`.

My only concern is related to the stability as none of this build helpers are marked as experimental and we are changing the flags that are currently used:
 * We are adding the runtime to `AutoToolsBuildEnvironment` (@SSE4, please include this in the changelog)
 * We are passing a lot of new flags to the `Meson` one

How does this play with existing recipes? Is this breaking existing behavior? Does it require to change existing recipes so they no longer use the _ugly workaround_? Can we consider it a _bug_ and just fix it?

**Comment by SSE4** [comment]
@jgsogo yes, we may change recipes to remove existing workarounds. right now we have to deal with compiler.runtime manually, e.g.:
https://github.com/bincrafters/conan-libiconv/blob/testing/1.15/conanfile.py#L91
https://github.com/bincrafters/conan-ffmpeg/blob/stable/4.1/conanfile.py#L255
otherwise, we don't get correct binaries. same story about meson:
https://github.com/bincrafters/conan-glib/blob/stable/2.58.3/conanfile.py#L79
https://github.com/bincrafters/conan-gstreamer/blob/stable/1.16.0/conanfile.py#L62
and it doesn't seem to be breaking - if recipe is already adding these flags, adding them for the second time will have no effect

**Comment by jgsogo** [comment]
...but if someone was not adding all those flags, now he will. Is this breaking, @lasote? Is it enough to add a paragraph in the release blog-post?

**Comment by SSE4** [comment]
if someone didn't add those flags, he got incorrect package. anyway, it's up to you to decide.

**Comment by jgsogo** [comment]
Ok, thanks, so let's go with it!

**Pull request merged**